File: docs/features/processing/_snippets/cpp/slicing.cpp
/*
 * Copyright (c) Ouster, Inc.
 * All rights reserved.
 *
 * Documentation snippets for the processing guide (C++ slicing examples).
 */

#include <cstdint>
#include <iostream>
#include <stdexcept>
#include <string>
#include <vector>

#include "ouster/core/frame_set_source.h"
#include "ouster/core/lidar_frame.h"
#include "ouster/core/open_source.h"
#include "ouster/core/xyzlut.h"

//! [doc-stag-slicing-imports]
using namespace ouster::sdk;
//! [doc-etag-slicing-imports]

namespace ouster {
namespace docs {

core::AnyFrameSetSource open_indexed_source_example_cpp(
    const std::string& pcap_path = "<SAMPLE_DATA_PCAP_PATH>") {
    // clang-format off
    //! [doc-stag-slicing-open-source]
    auto source = sdk::open_source(pcap_path, [](sdk::FrameSetSourceOptions& opts) { 
                                   opts.index = true; });
    //! [doc-etag-slicing-open-source]
    // clang-format on
    return source;
}

void print_nth_frame_example_cpp(const core::FrameSetSource& source) {
    if (!source.is_indexed()) {
        throw std::runtime_error("Random access requires an indexed FrameSetSource.");
    }
    // clang-format off
    //! [doc-stag-slicing-print-nth]
    auto frame_set = source[9];
    for (auto& frame : frame_set) {
        if (frame) {
            std::cout << frame->frame_id << '\n'; } }
    // To directly access the 9th frame from sensor index 0 without looping.
    auto frame = source[9][0];
    std::cout << frame->frame_id << '\n';
    //! [doc-etag-slicing-print-nth]
    // clang-format on
}

void print_last_frame_example_cpp(const core::FrameSetSource& source) {
    if (!source.is_indexed()) {
        throw std::runtime_error("Random access requires an indexed FrameSetSource.");
    }
    // clang-format off
    //! [doc-stag-slicing-print-last]
    auto frame = source[-1][0];
    std::cout << frame->frame_id << '\n';
    //! [doc-etag-slicing-print-last]
    // clang-format on
}

void print_first_n_frames_example_cpp(const core::FrameSetSource& source) {
    if (!source.is_indexed()) {
        throw std::runtime_error("Random access requires an indexed FrameSetSource.");
    }
    // clang-format off
    //! [doc-stag-slicing-first-n]
    for (const auto& frame_set : source[{0, 9}]) {
        for (auto& frame : frame_set) {
            if (frame) { 
                std::cout << frame->frame_id << '\n'; } } }
    //! [doc-etag-slicing-first-n]
    // clang-format on
}

void print_last_n_frames_example_cpp(const core::FrameSetSource& source) {
    if (!source.is_indexed()) {
        throw std::runtime_error("Random access requires an indexed FrameSetSource.");
    }
    // clang-format off
    //! [doc-stag-slicing-last-n]
    for (const auto& frame_set : source[{-9, -1}]) {
        for (const auto& frame : frame_set) {
            if (frame) { 
                std::cout << frame->frame_id << '\n'; } } }
    //! [doc-etag-slicing-last-n]
    // clang-format on
}

void print_step_sliced_frames_example_cpp(const core::FrameSetSource& source) {
    if (!source.is_indexed()) {
        throw std::runtime_error("Random access requires an indexed FrameSetSource.");
    }
    // clang-format off
    //! [doc-stag-slicing-step]
    // prints every second frame in the first ten
    for (const auto& frame_set : source[{0, 10, 2}]) {
        for (const auto& frame : frame_set) {
            if (frame) { std::cout << frame->frame_id << '\n'; } 
        } }
    //! [doc-etag-slicing-step]
    // clang-format on
}

void slicing_as_source_example_cpp(const core::FrameSetSource& source) {
    if (!source.is_indexed()) {
        throw std::runtime_error("Random access requires an indexed FrameSetSource.");
    }
    // clang-format off
    //! [doc-stag-slicing-subsource]
    auto source2 = source[{5, 10}];
    // This should print 5 since source2 is scoped to the range [5, 10]
    std::cout << "source2 length: " << source2.size() << '\n';
    // This is equivalent to calling source[5][0]
    const auto first_frame = source2[0][0];
    std::cout << first_frame->frame_id << '\n';
    // This is equivalent to calling source[9][0]
    const auto fifth_frame = source2[4][0];
    std::cout << fifth_frame->frame_id << '\n';

    // Use source2 as an iterator similar to the main source, assumes only one sensor
    for (const auto& frame_set : source2) {
        if (frame_set.size() > 0 && frame_set[0]) {
            std::cout << frame_set[0]->frame_id << '\n'; } }
    
    // it is possible to sub slice, meaning take the result of a previous slice operation and slice it
    // Thus, the following statement is valid
    auto source3 = source2[{2, 4}];
    std::cout << "source3 length: " << source3.size() << '\n';  // equivalent to source[7:9]
    //! [doc-etag-slicing-subsource]
    // clang-format on
}

void first_valid_column_example_cpp(const core::FrameSetSource& source) {
    if (!source.is_indexed()) {
        throw std::runtime_error("Random access requires an indexed FrameSetSource.");
    }
    // clang-format off
    //! [doc-stag-slicing-first-valid-column]
    auto frame = source[0][0]; 
    std::cout << frame->get_first_valid_column() << '\n';
    //! [doc-etag-slicing-first-valid-column]
    // clang-format on
}

void last_valid_column_example_cpp(const core::FrameSetSource& source) {
    if (!source.is_indexed()) {
        throw std::runtime_error("Random access requires an indexed FrameSetSource.");
    }
    // clang-format off
    //! [doc-stag-slicing-last-valid-column]
    auto frame = source[0][0]; 
    std::cout << frame->get_last_valid_column() << '\n';
    //! [doc-etag-slicing-last-valid-column]
    // clang-format on
}

}  // namespace docs
}  // namespace ouster

File: docs/features/processing/_snippets/python/slicing.py
"""Documentation snippets for FrameSetSource indexing and slicing."""
from __future__ import annotations

from ouster.sdk import core


def open_indexed_source_example(pcap_path: str = "<SAMPLE_DATA_PCAP_PATH>") -> core.FrameSetSource:
    """Doc example for enabling indexing when opening a PCAP/OSF source."""
    # [doc-stag-slicing-imports]
    from ouster import sdk
    # [doc-etag-slicing-imports]
    # [doc-stag-slicing-open-source]
    source = sdk.open_source(pcap_path,
                             index=True)
    # [doc-etag-slicing-open-source]
    return source


def print_nth_frame_example(source: core.FrameSetSource) -> None:
    """Doc example that prints the tenth frame's frame id."""
    # [doc-stag-slicing-print-nth]
    frame_set = source[9]
    for frame in frame_set:
        if frame is not None:
            print(frame.frame_id)
    # To directly access 9th frame of sensor with index 0.
    frame = source[9][0]
    assert frame is not None
    print(frame.frame_id)
    # [doc-etag-slicing-print-nth]


def print_last_frame_example(source: core.FrameSetSource) -> None:
    """Doc example that prints the last frame's frame id."""
    # [doc-stag-slicing-print-last]
    frame = source[-1][0]
    assert frame is not None
    print(frame.frame_id)
    # [doc-etag-slicing-print-last]


def print_first_n_frames_example(source: core.FrameSetSource) -> None:
    """Doc example that iterates over the first n frames."""
    # [doc-stag-slicing-first-n]
    for frame_set in source[0:9]:
        for frame in frame_set:
            if frame is not None:
                print(frame.frame_id)
    # [doc-etag-slicing-first-n]


def print_last_n_frames_example(source: core.FrameSetSource) -> None:
    """Doc example that iterates over the last nine frames."""
    # [doc-stag-slicing-last-n]
    for frame_set in source[-9:-1]:
        for frame in frame_set:
            if frame is not None:
                print(frame.frame_id)
    # [doc-etag-slicing-last-n]


def print_step_sliced_frames_example(source: core.FrameSetSource) -> None:
    """Doc example that demonstrates step slicing semantics."""
    # [doc-stag-slicing-step]
    # prints every second frame in the first ten
    for frame_set in source[0:10:2]:
        for frame in frame_set:
            if frame is not None:
                print(frame.frame_id)
    # [doc-etag-slicing-step]


def slicing_as_source_example(source: core.FrameSetSource) -> None:
    """Doc example that shows slicing returning another FrameSetSource."""
    # [doc-stag-slicing-subsource]
    source2 = source[5:10]
    # This should print 5 since source2 is scoped to the range [5, 10]
    print("source2 length:", len(source2))
    # This is equivalent to calling source[5][0]
    print(source2[0][0].frame_id)  # type: ignore[union-attr]

    # This is equivalent to calling source[9][0]
    print(source2[4][0].frame_id)  # type: ignore[union-attr]


    # Use source2 as an iterator similar to the main source, assumes only one sensor
    source_iter = iter(source2)
    for frame, in source_iter:
        assert frame is not None
        print(frame.frame_id)

    # it is possible to sub slice, meaning take the result of a previous slice operation and slice it
    # Thus, the following statement is valid
    source3 = source2[2:4]  # equivalent to source[7:9]
    print("source3 length:", len(source3))
    # [doc-etag-slicing-subsource]


def first_valid_column_example(source: core.FrameSetSource) -> None:
    """Print the first valid column index from the first frame."""
    # [doc-stag-slicing-first-valid-column]
    frame = source[0][0]
    assert frame is not None
    print(frame.get_first_valid_column())
    # [doc-etag-slicing-first-valid-column]


def last_valid_column_example(source: core.FrameSetSource) -> None:
    """Print the last valid column index from the first frame."""
    # [doc-stag-slicing-last-valid-column]
    frame = source[0][0]
    assert frame is not None
    print(frame.get_last_valid_column())
    # [doc-etag-slicing-last-valid-column]


if __name__ == "__main__":
    raise SystemExit("This module provides documentation snippets and is not executable.")
